File: src/vertical_control/scripts/real_sensor.py
#!/usr/bin/python
import cv2
import rospy
import numpy as np
from copy import deepcopy
from sensor_msgs.msg import Image
from threading import Lock
from std_msgs.msg import Float32
from std_msgs.msg import Float32MultiArray
from image_converter import ToOpenCV
from math import sqrt
from std_msgs.msg import Empty
from std_msgs.msg import Bool

class Vision_sensor():
	#class constants
	UPDATE_SPEED = 10 #process at 10 hz
	AREA_DIST_CONST = 35.623

	def __init__(self):
		
		self.latest_image = None

		self.image_lock = Lock()
		self.last = 0

		self.images = dict()
		self.height, self.width = (-1, -1)
		rospy.init_node('real_sensors', anonymous=False)
		self.hover_distance = rospy.get_param("v_controller/hover_distance")
		
		self.learner_pub = rospy.Publisher('v_controller/state', Float32MultiArray)
		self.not_visible_pub = rospy.Publisher('v_controller/visible', Bool)
		self.controller_pub = rospy.Publisher('v_controller/control_state', Float32MultiArray)

		self.video_sub = rospy.Subscriber('/ardrone/image_raw', Image, self.receive_image_callback)
		self.height_sub = rospy.Subscriber('/sonar_height', Range, self.receive_height_callback)
		self.rate = rospy.Rate(self.UPDATE_SPEED)

		

	def receive_image_callback(self, data):
		self.image_lock.acquire()
		try:
			self.latest_image = data 
		finally:
			self.image_lock.release()

	def receive_height_callback(self, Range_data):
		height = Range_data.range
		self.controller_pub.publish(None, [height])


	def sign(self, value):
		if (value > 0.0):
			return 2
		if (value < 0.0):
			return -2
		
		return 0


	def processing_function(self):
		print "waiting for images to come in..."
		while (not rospy.is_shutdown()) and self.latest_image is None:
			rospy.sleep(.5)		
		print "done!"

		self.get_image_size()

		while not rospy.is_shutdown(): #do image processing here
			#get the latest image
			self.image_lock.acquire()
			try:
				image = self.latest_image
			finally:
				self.image_lock.release()	

			image = np.asarray(ToOpenCV(image))

			x, y, distance = self.find_target(image)

			self.learner_pub.publish(None,[x,y])		
			self.controller_pub.publish(None, [distance])
			#need to add publisher for other info
			self.rate.sleep()
		
	def find_target (self, image):
		#parameters for locating green
		lower_target_range = (26, 74, 50)
		upper_target_range = (124, 240, 255)	

		x, y, distance = self.process_image (image, lower_target_range, upper_target_range)	 
		return x, y, distance

	def get_image_size(self):
		self.image_lock.acquire()
		try:
			image = self.latest_image
		finally:
			self.image_lock.release()	
		image = np.asarray(ToOpenCV(image))

		self.height, self.width, _ = image.shape

	def rescale(self, x, y):
		width_div = float(self.width)/2
		height_div = float(self.height)/2
		newx = float(x - width_div)/width_div			
		newy = float(y - height_div)/height_div			
		return newx,newy

	def process_image (self, image, lower_range, upper_range):
		#outputs = dict()
			
		#create copy of the image
		self.images["src"] = deepcopy(image) 
	
		#give it a nice blur
		#images["MFilter"] = cv2.medianBlur (images["Src"], 9)

		#convert to HSV
		self.images["hsv"] = cv2.cvtColor (self.images["src"], cv2.COLOR_BGR2HSV)

		#filter for color
		self.images["color_filter"] = cv2.inRange (self.images["hsv"], lower_range, upper_range)

		#locate contours
		contours, heirarchy = cv2.findContours (deepcopy (self.images["color_filter"]), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

		#self.images['contours'] = deepcopy(self.images['src'])
		#cv2.drawContours(self.images['contours'], contours, -1, (0,0,0), thickness=2)
			
		#self.images['center'] = deepcopy(self.images['src']) 
		#cv2.circle(self.images['center'], (self.width/2, self.height/2), 1, 0, 5)		
		#cv2.imshow('image', self.images['contours'])
		#cv2.waitKey(1)

		if contours: #check if we have any contours first
			contour = self.find_max_contour(contours)
			moment = cv2.moments(contour)
			area = moment["m00"] #cv2.contourArea(contours[0])
			print area
			if area > 800:
				xpos = moment["m10"] / area
				ypos = moment["m01"] / area
				distance = self.AREA_DIST_CONST * 1/sqrt(area)
				self.not_visible_pub.publish(1)
				
				xpos, ypos = self.rescale(xpos, ypos)
				self.last = self.sign(ypos)
				#print "a", area
				#print "d", distance
			else:
				xpos = 0
				ypos = self.last 
				distance = self.hover_distance
				self.not_visible_pub.publish(0)
		else:
			xpos = 0
			ypos = self.last
			distance = self.hover_distance
			self.not_visible_pub.publish(0)


		return xpos, ypos, distance 
	def find_max_contour(self, contours):
		max_contour = contours[0]
		max_area = cv2.moments(max_contour)["m00"]
		for contour in contours:
			if cv2.moments(contour)["m00"] > max_area:
				max_contour = contour
				max_area = cv2.moments(contour)["m00"]
		return max_contour
				
if __name__ == '__main__':
	sensor = Vision_sensor()	
	sensor.processing_function()

File: src/vertical_control/src/real_sensor.cpp
#include <ros/ros.h>
#include <ros/time.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp> // CV_AA

#include <math.h> // atan2
#include <sstream> // patch template to_string

#include <std_msgs/Float32.h>
#include <std_msgs/Float32MultiArray.h>
#include <std_msgs/Bool.h>
#include <std_msgs/Empty.h>
#include <ardrone_autonomy/Navdata.h>
#include <stdlib.h>


// default capture width and height
static const int FRAME_WIDTH = 640;
static const int FRAME_HEIGHT = 480;

// max number of objects to be detected in frame
static const int MAX_NUM_OBJECTS = 10; //50
// minimum and maximum object areai
static const int MIN_OBJECT_AREA = 50; // 5*5
static const int MAX_OBJECT_AREA = FRAME_HEIGHT*FRAME_WIDTH/1.5;

static const int UPDATE_SPEED = 200; // process at 10hz
static const float AREA_DIST_CONST = 35.623f;

static const float PI = 3.1415;

class VisionSensor {

    ros::Publisher learner_pub;
    ros::Publisher not_visible_pub;
    ros::Publisher controller_pub;
    ros::Subscriber height_sub;

    image_transport::ImageTransport it_;
    image_transport::Subscriber video_sub;
    //image_transport::Publisher image_pub_;

    cv_bridge::CvImagePtr latest_image;

    float hover_distance;

    int h_min_;
    int h_max_;
    int s_min_;
    int s_max_;
    int v_min_;
    int v_max_;

    float height;
    float width;
    float last;

    bool is_first_img;

public:
    VisionSensor(ros::NodeHandle &nh_):it_(nh_)
    {
        is_first_img = true;
        //latest_image = 0;
        // green
        h_min_ = 26;
        h_max_ = 124;
        s_min_ = 74;
        s_max_ = 240;
        v_min_ = 50;
        v_max_ = 255;
        // orange
        //h_min_ = 0;
        //h_max_ = 200;
        //s_min_ = 170;
        //s_max_ = 255;
        //v_min_ = 0;
        //v_max_ = 255;

        height = width = -1.0f;
        last = 0.0f;
        hover_distance = 2.0f;

        //nh_.getParam("v_controller/hover_distance", hover_distance);

        learner_pub = nh_.advertise<std_msgs::Float32MultiArray>("v_controller/state", 2);
        not_visible_pub = nh_.advertise<std_msgs::Bool>("v_controller/visible", 1);
        controller_pub = nh_.advertise<std_msgs::Float32MultiArray>("v_controller/control_state", 1);

        video_sub = it_.subscribe("/ardrone/image_raw", 1, &VisionSensor::receive_image_cb, this);
        height_sub = nh_.subscribe("/ardrone/navdata", 1, &VisionSensor::receive_height_cb, this);

        ros::spinOnce();
    }

    bool isFirstImgLoaded()
    {
        return is_first_img;
    }

    void start()
    {
        ros::Rate loop_rate(UPDATE_SPEED);

        while (ros::ok())
        {
            if (!is_first_img)
                processing_function();
            ros::spinOnce();
            loop_rate.sleep();
        }
    }

    void receive_image_cb(const sensor_msgs::ImageConstPtr& msg)
    {
        try
        {
            // OpenCV expects color images to use BGR channel order
            latest_image = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);

            if (is_first_img) {
                height = latest_image->image.size().height;
                width = latest_image->image.size().width;
                is_first_img = false;
            }
        }
        catch (cv_bridge::Exception& e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }
    }

    void receive_height_cb(const ardrone_autonomy::Navdata& msg)
    {
        float h = float(msg.altd) / 1000.0f;
        std_msgs::Float32MultiArray hArray;
        hArray.data.push_back(h);
        controller_pub.publish(hArray);
    }

    float sign(float value)
    {
        if (value > 0.0f) return 2.0f;
        if (value < 0.0f) return -2.0f;
        return 0.0f;
    }

    void processing_function()
    {
        ros::WallTime startt = ros::WallTime::now();
        //while (true)
        //    if (ros::ok() && latest_image != 0) break;

        float x = 0.0f, y = 0.0f, distance = 0.0f;
        find_target(latest_image, x, y, distance);

        std_msgs::Float32MultiArray pose;
        pose.data.push_back(x);
        pose.data.push_back(y);

        learner_pub.publish(pose);

        ros::WallTime endt = ros::WallTime::now();
        ros::WallDuration diff = endt - startt;
        ROS_DEBUG_STREAM_NAMED("vertical_control", "Object Detection duration " << diff.toSec());
    }

    void find_target(cv_bridge::CvImagePtr img, float &x, float &y, float &distance)
    {
        // Convert to HSV
        cv::Mat hsv;
        cv::cvtColor(img->image, hsv, CV_BGR2HSV);

        // Use smaller image for faster processing
        cv::Mat hsv_small(cv::Size(160,120), hsv.type());
        cv::resize(hsv, hsv_small, hsv_small.size());
 
        // filter for color
        cv::Mat color_filter;
        cv::inRange(hsv_small, cv::Scalar(h_min_, s_min_, v_min_), cv::Scalar(h_max_, s_max_, v_max_), color_filter);

        // eliminate noise
        morphOps(color_filter);

        std::string text;
        trackFilteredObject(color_filter, x, y, distance, text);

        cv::putText(img->image, text, cv::Point(0,50), 2, 1, cv::Scalar(0,255,0), 2);
        cv::imshow("threshold", color_filter);
        cv::imshow("frame", img->image);
        cv::waitKey(5);
    }


    void trackFilteredObject(cv::Mat &threshold, float &x, float &y, float &distance, std::string &text)
    {
        cv::Mat temp = threshold.clone();
        bool object_found = false;
        std_msgs::Bool b;
        float ref_area = 0.0f;

        // find contours of filtered image using openCV findContours functions
        // two vectors needed for output of findContours
        cv::vector< cv::vector<cv::Point> > contours;
        cv::vector<cv::Vec4i> hierarchy;
        cv::findContours(temp, contours, hierarchy, CV_RETR_CCOMP, CV_CHAIN_APPROX_SIMPLE);
        // use moments method to find our filtered object
        if (hierarchy.size() > 0)
        {
            int numObjects = hierarchy.size();
            // if number of objects greater than MAX_NUM_OBJECTS we have a noisy filter
            if (numObjects <= MAX_NUM_OBJECTS)
            {
                //int x = 0, y = 0;
                for (int index = 0; index >= 0; index = hierarchy[index][0])
                {
                    cv::Moments moment = cv::moments((cv::Mat)contours[index]);
                    double area = moment.m00;
                    //double area = cv::contourArea(contours[index]);

                    // if the area is less than 20px by 20px then it is probably just noise
                    // if the area is the same as the 3/2 of the image size, probably just a bad filter
                    // we only want the object with the largest area so we safe a reference area each
                    // iteration and compare it to the area in the next iteration.
                    if (area > MIN_OBJECT_AREA && area < MAX_OBJECT_AREA && area > ref_area)
                    //if (area > ref_area)
                    {
                        x = moment.m10/area;
                        y = moment.m01/area;
                        object_found = true;
                        ref_area = area;
                    }
                }
                // let user know you found an object
                if (object_found)
                {
                    float px = x / threshold.size().width;
                    float py = y / threshold.size().height;
                    x = ceil(width * px) - 12.0f;
                    y = ceil(height * py) - 12.0f;
                    distance = AREA_DIST_CONST * 1.0f / sqrt(ref_area);
                    b.data = 1;
                    not_visible_pub.publish(b);
                    rescale(x, y);
                    last = sign(y);
                    text = "Largest Contour found";
                    return; // ends function when object is found
                    //cv::putText(frame, "Tracking Object", cv::Point(0,50), 2, 1, cv::Scalar(0,255,0), 2);
                }
            }
        }

        // object not found
        x = 0.0f;
        y = last;
        distance = hover_distance;
        b.data = 0;
        not_visible_pub.publish(b);
        text = "Nothing found";
    }

    void rescale(float &x, float &y)
    {
        float width_div = float(width) / 4.0f;
        float height_div = float(height) / 3.0f;
        x = (x - width/2) / width_div;
        y = (y - height/2) / height_div;
    }

    void morphOps(cv::Mat &threshold)
    {
      // create structuring element that will be used to dilate and erode image
      // the element chosen is a 3px by 3px rectangle
      cv::Mat erode_kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2,2));
      // dilate with larger element so make sure object is nicely visible
      cv::Mat dilate_kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2,2)); //8,8
      cv::Mat dilate_kernel4 = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(4,4)); //8,8

      cv::erode(threshold, threshold, erode_kernel);
      cv::erode(threshold, threshold, erode_kernel);
      cv::erode(threshold, threshold, erode_kernel);
      cv::dilate(threshold, threshold, dilate_kernel);
      cv::dilate(threshold, threshold, dilate_kernel4);

    }

};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "real_sensors");
    ros::NodeHandle nh;

    VisionSensor vs(nh);
    //vs.start();

    ros::Rate loop_rate(UPDATE_SPEED);
    
    while (ros::ok())
    {
        if (!vs.isFirstImgLoaded())
            vs.processing_function();
        ros::spinOnce();
        loop_rate.sleep();
    }

    return 0;
}
